typing/typemod.ml

```diff
@@ -344,28 +346,60 @@ let check_usage_of_path_of_substituted_item paths ~loc ~lid env super =
+    application or recursive module). If it is the case, module aliases to [X]
+    or any suffix [X.X'] would become invalid. This function checks that there
+    are no module aliases to suffixes of the list of affected paths [paths]. The
+    [invalid_alias_path] argument is just used for the error message *)
+let check_invalid_aliases paths ~loc env super invalid_alias_path =
+  let would_become_invalid_path aliased_path =
+    List.exists
+      (fun p -> path_is_prefix aliased_path ~prefix:p ~strict:false) paths
   in
-  (* The last item is the one that's removed. We don't need to check how
-        it's used since it's replaced by a more specific type/module. *)
-  assert (match last with Pident _ -> true | _ -> false);
-  let iterator = match rest with
-    | [] -> iterator
-    | _ :: _ ->
-        check_usage_of_path_of_substituted_item rest ~loc ~lid env iterator
-  in
-  iterator.Btype.it_signature iterator sg
-  end
+  { super with
+    Btype.it_signature_item = (fun self -> function
+        | Sig_module (id, _, {md_type = Mty_alias aliased_path}, _, _)
+          when would_become_invalid_path aliased_path ->
+            raise(Error(loc, Lazy.force !env,
+                        With_creates_invalid_aliases
+                          (id, aliased_path, invalid_alias_path)))
+        | sig_item ->
+            super.Btype.it_signature_item self sig_item
+      );
+  }
 
-let check_usage_after_substitution env ~loc ~lid paths sg =
-  match paths with
-  | [_] -> ()
-  | _ -> do_check_after_substitution env ~loc ~lid paths sg
+(** This function checks the effect of destructive substitutions and the
+    introduction of invalid aliases *)
+let check_usage_after_substitution env ~loc ~lid paths ?(invalid_alias=None) sg=
+  match paths, invalid_alias with
+  (* Shallow substitution without introduction of an invalid alias *)
+  | [_], None -> ()
+  (* Deep substitution, or possible introduction of invalid alias *)
+  | _, _ ->
+      with_type_mark begin fun mark ->
+        let env, base_iterator =
+          iterator_with_env (Btype.type_iterators mark) env in
+        let last, rest = match List.rev paths with
+          | [] -> assert false (* affected paths must be a non-empty list *)
+          | last :: rest -> last, rest
+        in
+        (* The last item is the one that's removed. We don't need to check how
+              it's used since it's replaced by a more specific type/module. *)
+        assert (match last with Pident _ -> true | _ -> false);
+        let iterator = match rest with
+          | [] -> base_iterator
+          | _ :: _ ->
+              check_usage_of_path_of_substituted_item rest ~loc ~lid env
+                base_iterator
+        in
+        let iterator' = match invalid_alias with
+          | None -> iterator
+          | Some invalid_alias_path ->
+              check_invalid_aliases paths ~loc env iterator invalid_alias_path
+        in
+        iterator.Btype.it_signature iterator' sg
```

Comment:
I am wondering if we want to move the testing of conditions to `check_usage_of_path_of_substituted_item` and `check_invalid_aliases` in order to reduce the combination of the two iterator transformer to  just
```ocaml
let iterator = 
       base_iterator 
  |> check_usage_of_path_of_substituted_item paths ~loc ~lid env
  |> check_invalid_aliases paths ~loc env iterator invalid_alias_path
in
iterator.Btype.it_signature iterator' sg
```